core/src/com/unciv/logic/BarbarianManager.kt

```diff
@@ -0,0 +1,265 @@
+package com.unciv.logic
+
+import com.badlogic.gdx.math.Vector2
+import com.unciv.Constants
+import com.unciv.logic.civilization.NotificationIcon
+import com.unciv.logic.map.TileInfo
+import com.unciv.logic.map.TileMap
+import com.unciv.models.metadata.GameSpeed
+import java.util.*
+import kotlin.math.min
+import kotlin.math.pow
+
+class BarbarianManager {
+    val camps = ArrayList<Encampment>()
+
+    @Transient
+    lateinit var gameInfo: GameInfo
+
+    @Transient
+    lateinit var tileMap: TileMap
+
+    fun clone(): BarbarianManager {
+        val toReturn = BarbarianManager()
+        toReturn.camps.addAll(camps.map { it.clone() } )
+        return toReturn
+    }
+
+    fun setTransients(gameInfo: GameInfo) {
+        this.gameInfo = gameInfo
+        this.tileMap = gameInfo.tileMap
+
+        // Add any preexisting camps as Encampment objects
+        for (tile in tileMap.values) {
+            if (tile.improvement == Constants.barbarianEncampment
+                && camps.none { it.position == tile.position }) {
```

Comment:
faster if camps where a HashMap. Vector2 obeys the equality contract and is safe to use as Hash key.